File: packages/guards/src/pii-pl.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

type PiiPlEntity = 'pesel' | 'nip' | 'regon' | 'dowod' | 'phone-pl';
interface PiiPlOptions { entities: PiiPlEntity[]; action: 'block' | 'warn' | 'mask'; }
interface PiiPlMatch { type: PiiPlEntity; value: string; start: number; end: number; }

const PATTERNS: Record<PiiPlEntity, RegExp> = {
  pesel: /\b\d{11}\b/g,
  nip: /\b\d{3}-\d{3}-\d{2}-\d{2}\b/g,
  regon: /\b\d{9}(?:\d{5})?\b/g,
  dowod: /\b[A-Z]{3}\d{6}\b/g,
  'phone-pl': /\b(?:\+?48[-.\s]?)?\d{3}[-.\s]?\d{3}[-.\s]?\d{3}\b/g,
};
const MASK: Record<PiiPlEntity, string> = { pesel: '[PESEL]', nip: '[NIP]', regon: '[REGON]', dowod: '[Dowod]', 'phone-pl': '[Telefon]' };

function detect(text: string, entities: PiiPlEntity[]): PiiPlMatch[] {
  const m: PiiPlMatch[] = [];
  for (const e of entities) { const re = new RegExp(PATTERNS[e].source, 'g'); let x; while ((x = re.exec(text))) m.push({ type: e, value: x[0], start: x.index, end: x.index + x[0].length }); }
  return m.sort((a, b) => b.start - a.start);
}

export function piiPl(options: PiiPlOptions): Guard {
  return { name: 'pii-pl', version: '0.1.0', description: 'Polish PII (PESEL, NIP, REGON, dowod, phone)', category: 'privacy', supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now(); const matches = detect(text, options.entities); const triggered = matches.length > 0;
      if (!triggered) return { guardName: 'pii-pl', passed: true, action: 'allow', latencyMs: Math.round(performance.now() - start) };
      if (options.action === 'mask') { let r = text; for (const m of matches) r = r.slice(0, m.start) + MASK[m.type] + r.slice(m.end); return { guardName: 'pii-pl', passed: true, action: 'override', overrideText: r, latencyMs: Math.round(performance.now() - start), details: { detected: matches.map(({ type }) => ({ type })) } }; }
      return { guardName: 'pii-pl', passed: false, action: options.action, message: `Polish PII: ${[...new Set(matches.map(m => m.type))].join(', ')}`, latencyMs: Math.round(performance.now() - start), details: { detected: matches.map(({ type, value }) => ({ type, value })) } };
    },
  };
}

File: python/open_guardrail/guards/pii_pl.py
"""Polish PII detection and masking guard."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PATTERNS: dict[str, re.Pattern[str]] = {
    "pesel": re.compile(r"\b\d{11}\b"),
    "nip": re.compile(r"\b\d{3}-?\d{3}-?\d{2}-?\d{2}\b"),
    "regon": re.compile(r"\b\d{9}(?:\d{5})?\b"),
    "dowod": re.compile(r"\b[A-Z]{3}\d{6}\b"),
    "phone-pl": re.compile(r"\b(?:\+48|0048)\s?\d{3}[\s\-]?\d{3}[\s\-]?\d{3}\b"),
}

_MASK_LABELS = {"pesel": "[PESEL]", "nip": "[NIP]", "regon": "[REGON]", "dowod": "[DOWÓD]", "phone-pl": "[TELEFON-PL]"}


def _detect(text: str, entities: list[str]) -> list[dict]:
    matches = []
    for e in entities:
        pat = _PATTERNS.get(e)
        if pat:
            for m in pat.finditer(text):
                matches.append({"type": e, "value": m.group(), "start": m.start(), "end": m.end()})
    matches.sort(key=lambda x: x["start"], reverse=True)
    return matches


def _mask(text: str, matches: list[dict]) -> str:
    r = text
    for m in matches:
        r = r[: m["start"]] + _MASK_LABELS.get(m["type"], "[PII]") + r[m["end"] :]
    return r


class _PiiPl:
    def __init__(self, *, action: str = "block", entities: Optional[List[str]] = None) -> None:
        self.name = "pii-pl"
        self.action = action
        self._entities = entities or list(_PATTERNS.keys())

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matches = _detect(text, self._entities)
        elapsed = (time.perf_counter() - start) * 1000
        if not matches:
            return GuardResult(guard_name="pii-pl", passed=True, action="allow", latency_ms=round(elapsed, 2))
        detected = [{"type": m["type"], "value": m["value"]} for m in matches]
        if self.action == "mask":
            return GuardResult(guard_name="pii-pl", passed=True, action="override", override_text=_mask(text, matches), latency_ms=round(elapsed, 2), details={"detected": detected})
        return GuardResult(guard_name="pii-pl", passed=False, action=self.action, message="Polish PII detected", latency_ms=round(elapsed, 2), details={"detected": detected})


def pii_pl(*, action: str = "block", entities: Optional[List[str]] = None) -> _PiiPl:
    return _PiiPl(action=action, entities=entities)
